### DISCUSSION TITLE
Fluent API to await multiple calls and get their respective results

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#102] [General]
# Fluent API to await multiple calls and get their respective results

Sometimes, you have multiple async calls to make, and you want to do that asynchronously and get the results afterward. Let's build a fluent API to do that.

https://steven-giesel.com/blogPost/d55cdae5-7294-40cc-a9d5-7552e092205e

**Comment by Ghost4Man** [comment]
An alternative option is to use the [TaskTupleAwaiter](https://www.nuget.org/packages/TaskTupleAwaiter/) package (or create a custom `GetAwaiter` extension method on tuples of Tasks of various arities), which can be used like this:

```cs
var (name, description) = await (GetNameAsync(), GetDescriptionAsync());
```

**Reply by linkdotnet** [reply]
That is a much better option as this is just an educational blogpost meant to be refined :)

**Comment by toupswork** [comment]
Hey. I was just reading about using System.Tasks.DataFlow, and it made me think about this.  Looks like DataFlow has great advantages for orchestrating things like a series of API calls or DB queries containing a mixture of calls that must be done sequentially as well as calls that can be done in parallel. have you heard of this data flow API before?

**Reply by linkdotnet** [reply]
No - that one is new for me. It reminds me of the API Visual Studio is offering for Tasks.
Not sure if I have a direct need for that, bt good to know anyway.

**Reply by linkdotnet** [reply]
Did you find any use case for yourself?

**Reply by toupswork** [reply]
Yeah. Most all of my C# coding is APIs which aggregate data from multiple, other APIs.  So going forward, rather than making a series of await calls along with implicit parallelism using the `Task.WhenAll()`, I'll be using Data Flows

**Reply by toupswork** [reply]
I learned something else I thought I'd share. You may already know this, but just in case, if you want to take a string and get a substring of it without creating a new heap allocation for the substring:
1. Convert the string to a `ReadOnlySpan<char>`
2. Perform a slice to get the substring
3. Convert the sliced span back to a string
The excised chars from the substring are re-used rather than allocating them on heap.

Here's another one. If you want to use a struct that implements an interface but you don't want it to be boxed, have the method returning the struct take in a generic argument and add a constraint on it that it must extend the interface. 

And one more, the JIT is smart enough so that if you create a closure using a lambda, it will actually allocate the class that holds the closure on the stack.

**Reply by linkdotnet** [reply]
Yeah I used that in my `ValueStringBuilder` library quite a bit.

> And one more, the JIT is smart enough so that if you create a closure using a lambda, it will actually allocate the class that holds the closure on the stack.

That is interesting. Do you know by any chance what are the requirements for this? When does the JIT lean towards this optimization?

**Reply by toupswork** [reply]
Yes, the JIT compiler in .NET has the capability to optimize certain allocations by placing them on the stack, This optimization is primarily done for closures in lambda expressions or local functions. However, it doesn't always occur and depends on a few specific conditions:

## Requirements for the JIT to Allocate Closures on the Stack:
1. **Closure does not escape the method context:**
The most important factor is that the closure (i.e., the class that holds the captured variables) must not "escape" the method where it was created. In other words, the JIT can only allocate the closure on the stack if it's certain that the closure will not be used after the method has returned. If the closure is passed to another method or stored in a longer-lived variable (e.g., a field or heap-allocated object), the JIT cannot place it on the stack.

2. **Lambda or local function is not assigned to a delegate:**
If a lambda or local function is assigned to a delegate that may outlive the method scope (e.g., by storing it in a field or passing it to asynchronous methods or event handlers), the closure must be heap-allocated to ensure its lifetime is not tied to the method stack frame.

3. **No capturing of heap-allocated references:**
If the lambda captures references to objects that are allocated on the heap, the closure will likely also be heap-allocated to avoid lifetime issues. The JIT cannot guarantee that the referenced object will still be valid if the lambda outlives the stack frame.

4. **Inlining considerations:**
In some cases, if the lambda expression or local function is small and the JIT decides to inline the enclosing method, it can further optimize the closure handling by avoiding heap allocation and keeping everything on the stack.

### When Does the JIT Favor This Optimization?
The JIT tends to favor stack allocation for closures when:

* The closure is short-lived and does not outlive the method call.
* The closure captures only local variables or arguments that are also stack-allocated.
* No complex control flow (e.g., async/await, thread switching) that would extend the closure's lifetime beyond the method.
#### Example:
Consider the following scenario:

```
void SomeMethod()
{
    int localVar = 42;
    Action action = () => Console.WriteLine(localVar);
    action();
}
```
In this case:

* The closure captures the `localVar` variable.
* The closure does not escape the scope of SomeMethod, as the lambda is invoked within the same method without being passed to other methods or stored elsewhere.
* The JIT can allocate the closure on the stack.

However, if you store the lambda in a field or return it from the method, the closure would have to be heap-allocated because its lifetime extends beyond the stack frame:

```
Action storedAction;

void SomeMethod()
{
    int localVar = 42;
    storedAction = () => Console.WriteLine(localVar); // Closure must be heap-allocated
}
```

**Reply by linkdotnet** [reply]
Interesting - I wonder if the same applies to a primitive array and LINQ queries. If I have something like:
```csharp
int[] myArray = GetValuesFromSomewhere();
return myArray.Where(s => s % 2 == 0).ToArray();
```
Not sure, if the JIT does know that those values wouldn't escape to the heap.

**Reply by toupswork** [reply]
Take a look at the JIT compiled code

**Reply by linkdotnet** [reply]
Sharplab.io - I forgot about that :)

**Reply by toupswork** [reply]
Hey, I just tried that example code you posted. I didn't even need to look at the x64 assembly; the IL op code says it's pushing it on the stack
![image](https://github.com/user-attachments/assets/ef62a186-a115-4c7d-9af1-52ea9d0fde71)

**Comment by toupswork** [comment]
I created the following extension methods for Linq to prevent closures. But it's good to know that in some cases these are not needed if the compiler generated class for the closure can be put on the stack. These are a naive implementation, as I know that the Microsoft linq extensions have more performant algorithms that, for example, check what the underlying enumerable is to route it to a more optimized code for it.

```csharp
using System.Diagnostics;

namespace System.Linq;

public static partial class Enumerable
{

  [DebuggerStepThrough, DebuggerNonUserCode]
  public static IEnumerable<TResult> SelectWithContext<TSource, TContext, TResult>(this IEnumerable<TSource> @this,
                                                                                   TContext context,
                                                                                   Func<TContext, TSource, TResult> selector)
  {
    foreach (var item in @this)
    {
      yield return selector(context, item);
    }
  }
  [DebuggerStepThrough, DebuggerNonUserCode]
  public static IEnumerable<TSource> WhereWithContext<TSource, TContext>(
    this IEnumerable<TSource> @this, TContext context, Func<TContext, TSource, bool> predicate)
  {
    foreach (var item in @this)
    {
      if (predicate(context, item))
      {
        yield return item;
      }
    }
  }
}
```

**Reply by linkdotnet** [reply]
It would be really nice, if you could force the passed in lambda to be static. Sure you could do this:
```csharp
blub.WhereWithContext(static (...) => {});
```

But people don't have to use `static`.

**Reply by toupswork** [reply]
Yeah. I know you already know that putting static in front of it doesn't actually change the underlying IL generated and that it's only to give you a warning if you try to capture a closure
